### PR TITLE
[ISSUE-732] Fixed stacked ACR with LVG Volumes on reservation stage

### LINKED ISSUE #732: ACR with LVG Volumes stacked in reservation controller
**Describe the bug**
Found delays between CreateVolume requests coming to CSI Controller:

1. 2 minutes:
```
./csi-baremetal-controller/csi-baremetal-controller-6c89866b67-9d64j/controller.log:Jan 28 04:35:28.082 [INFO] [CSIControllerService] [CreateVolume] [pvc-8f7e09bf-cc83-4791-8111-87978dcd1839] Waiting until volume will reach Created status. Current status - CREATING
...
./csi-baremetal-controller/csi-baremetal-controller-6c89866b67-9d64j/controller.log:Jan 28 04:38:25.778 [INFO] [CSIControllerService] [CreateVolume] [pvc-cd14889e-38d4-4335-a890-562e86a98ee2] Waiting until volume will reach Created status. Current status - CREATING
```
2. 5 minutes:
```
./csi-baremetal-controller/csi-baremetal-controller-6c89866b67-9d64j/controller.log:Jan 28 04:38:25.858 [INFO] [CSIControllerService] [CreateVolume] [pvc-a922fcb9-f6d0-4b7e-9722-58e50b8fc855] Waiting until volume will reach Created status. Current status - CREATING
...
./csi-baremetal-controller/csi-baremetal-controller-6c89866b67-9d64j/controller.log:Jan 28 04:43:55.675 [INFO] [CSIControllerService] [CreateVolume] [pvc-b1a506f1-5bbf-4003-bcee-36da17efdf91] Waiting until volume will reach Created status. Current status - CREATING
```
3. 11 minutes:
```
./csi-baremetal-controller/csi-baremetal-controller-6c89866b67-9d64j/controller.log:Jan 28 04:43:55.732 [INFO] [CSIControllerService] [CreateVolume] [pvc-b6c3a2a1-10ce-41a1-a1a8-7d9409d7b787] Waiting until volume will reach Created status. Current status - CREATING
...
./csi-baremetal-controller/csi-baremetal-controller-6c89866b67-9d64j/controller.log:Jan 28 04:54:25.713 [INFO] [CSIControllerService] [CreateVolume] [pvc-0aa1a5b9-0796-4dcd-87f6-4de5d0cfbb95] Waiting until volume will reach Created status. Current status - CREATING
```
4. 17 minutes:
```
./csi-baremetal-controller/csi-baremetal-controller-6c89866b67-9d64j/controller.log:Jan 28 04:54:25.765 [INFO] [CSIControllerService] [CreateVolume] [pvc-622bc5cf-e65d-4675-b692-c06aed2db901] Waiting until volume will reach Created status. Current status - CREATING
./csi-baremetal-controller/csi-baremetal-controller-6c89866b67-9d64j/controller.log:Jan 28 05:11:32.102 [INFO] [CSIControllerService] [CreateVolume] [pvc-2db5b139-415c-4e68-8961-725fdf252d56] Waiting until volume will reach Created status. Current status - CREATING
```
Checked first delay:

   - Previous ACR was removed
```
Jan 28 04:35:28.082 [INFO] [VolumeOperationsImpl] [ReservationHelper.removeACR] [pvc-8f7e09bf-cc83-4791-8111-87978dcd1839] ACR default-sts-pvc-hdd-hddlvg-test-4-1 removed
```
   - Scheduler extender getting filter requests
```
./csi-baremetal-se/csi-baremetal-se-v5r6j/scheduler-extender.log:Jan 28 04:35:28.100 [INFO] [Extender] [FilterHandler] [sts-pvc-hdd-hddlvg-test-0-4] [6623647c-480c-45eb-991a-ae2ccbd83524] Construct response. Get 2 nodes in request. Among them suitable nodes count is 0. Filtered out nodes - map[]
./csi-baremetal-se/csi-baremetal-se-v5r6j/scheduler-extender.log:Jan 28 04:35:39.071 [INFO] [Extender] [FilterHandler] [sts-pvc-hdd-hddlvg-test-0-4] [b1635317-fbf6-4872-a250-67edd04d5ef0] Filtering
./csi-baremetal-se/csi-baremetal-se-v5r6j/scheduler-extender.log:Jan 28 04:35:39.071 [DEBU] [Extender] [gatherCapacityRequestsByProvisioner] [sts-pvc-hdd-hddlvg-test-0-4]
```
 - Controller started to handle reservation requests with 2 minutes delay
```
Jan 28 04:37:46.969 [INFO] [ReservationController] [Reconcile] [default-sts-pvc-hdd-hddlvg-test-0-4] Reservation status: REQUESTED
```
**Environment (please complete the following information):**
 - Platform: Openshift

**To Reproduce**
Create 30+ pods with LVG Volumes

### PR DESCRIPTION
**Pull request by safronovD**
### Resolves #732 

**RCA**
Reservation controller process only one ACR with LVG Volumes for a time. So many reconciliation attempts on the reservation stage are expected. Default RateLimiter in reconcile-controller is exponential, which leads to increasing timeout after 100 retries to 5+ mins.

- Add linear RateLimiter
- Pass RateLimiter parameters from csi-deployment chart

## PR checklist
- [x] Add link to the issue
- [x] Choose Project
- [x] Choose PR label
- [ ] New unit tests added
- [ ] Modified code has meaningful comments
- [ ] All TODOs are linked with the issues
- [ ] All comments are resolved

## Testing
_Provide test details_

### CODE REVIEW

File: pkg/crcontrollers/reservation/reservationcontroller.go [hunk 1]
```diff
@@ -43,6 +45,17 @@ func NewController(client *k8s.KubeClient, log *logrus.Logger, sequentialLVGRese
 func (c *Controller) SetupWithManager(mgr ctrl.Manager) error {
 	return ctrl.NewControllerManagedBy(mgr).
 		For(&acrcrd.AvailableCapacityReservation{}).
+		WithOptions(controller.Options{
+			// Rater controls timeout between reconcile attempts (if result.Requeue is true)
+			// Default RateLimiter is exponential, which leads to increasing timeout after 100 retries to 5+ mins
+			// Many attempts are expected during processing ACRs with LVG Volumes, so we need to use linear RateLimiter
+			// Here we make first 30 retries with 1.5 sec timeout and then it will be 12 sec forever
+			RateLimiter: workqueue.NewItemFastSlowRateLimiter(
```
Comment (rkiselenko) [inline pkg/crcontrollers/reservation/reservationcontroller.go:53]: Is this applicable for all `ctrl.Result{Requeue: true}` attempts ? Maybe it's better to use `ctrl.Result{RequeueAfter: time.Duration}` for processing ACRs with LVG Volumes ?

Comment (safronovD) [reply pkg/crcontrollers/reservation/reservationcontroller.go:53]: Yes, timeouts would be the same for all errors. We have not so many endpoints in this controller and these changes won't affect them in the short attempts queue.
If we use `ctrl.Result{RequeueAfter: time.Duration}`, timeout will be const on all intervals. It is not good due to many calls to k8s API

**Review by rkiselenko — APPROVED** [review]

File: pkg/crcontrollers/reservation/reservationcontroller.go [hunk 2]
```diff
@@ -43,6 +45,17 @@ func NewController(client *k8s.KubeClient, log *logrus.Logger, sequentialLVGRese
...
+				1500*time.Millisecond, // fastTimeout
```
Comment (mishoyama) [inline pkg/crcontrollers/reservation/reservationcontroller.go:54]: I think these parameters must be configurable - they depend on number of volumes and replicas

Comment (safronovD) [reply pkg/crcontrollers/reservation/reservationcontroller.go]: Fixed - https://github.com/dell/csi-baremetal-operator/pull/111

File: pkg/crcontrollers/reservation/reservationcontroller.go [hunk 3]
```diff
@@ -19,30 +23,55 @@ import (
...
+	defaulSlowTimeout = 12 * time.Second
+	defaulAttemptsNum = 30
 )
 
 // Controller to reconcile aviliablecapacityreservation custom resource
 type Controller struct {
 	client                 *k8s.KubeClient
 	log                    *logrus.Entry
 	capacityManagerBuilder capacityplanner.CapacityManagerBuilder
+	fastTimeout            time.Duration
```
Comment (rkiselenko) [inline pkg/crcontrollers/reservation/reservationcontroller.go:42]: The variables should have name like this:
```golang
fastTimeoutMs   = 1500 * time.Millisecond
slowTimeoutSec = 12 * time.Second
```
So the type of duration denoted in the name.

File: pkg/crcontrollers/reservation/reservationcontroller.go [hunk 4]
```diff
@@ -19,30 +23,55 @@ import (
...
+	slowTimeout            time.Duration
+	maxAttempts            uint64
```
Comment (rkiselenko) [inline pkg/crcontrollers/reservation/reservationcontroller.go:44]: Why `uint64` ? Use `int`

File: pkg/crcontrollers/reservation/reservationcontroller.go [hunk 5]
```diff
@@ -135,3 +164,39 @@ func (c *Controller) handleReservationUpdate(ctx context.Context, log *logrus.En
...
+		slowTimeoutStr = os.Getenv(slowTimeoutEnv)
+		maxAttemptsStr = os.Getenv(attemptsNumEnv)
+
+		fastTimeout time.Duration
+		slowTimeout time.Duration
+		maxAttempts uint64
+		err         error
+	)
+
+	fastTimeout, err = time.ParseDuration(fastTimeoutStr)
```
Comment (rkiselenko) [inline pkg/crcontrollers/reservation/reservationcontroller.go:180]: `0s` and `0` also parsable and won't raise an error. The below code too.

**Review by rkiselenko — CHANGES_REQUESTED** [review]

File: pkg/crcontrollers/reservation/reservationcontroller.go (hunk 4 above)
Comment (mishoyama) [reply pkg/crcontrollers/reservation/reservationcontroller.go:44]: @rkiselenko `maxAttempts` cannot be negative

Comment (rkiselenko) [reply pkg/crcontrollers/reservation/reservationcontroller.go:44]: This validation should be on the `CRD` level, so negative number wont pass to controller. https://book.kubebuilder.io/reference/markers/crd-validation.html

File: pkg/crcontrollers/reservation/reservationcontroller.go (hunk 3 above)
Comment (safronovD) [reply pkg/crcontrollers/reservation/reservationcontroller.go]: Refactored with RateLimiter args names from [go-client](https://github.com/kubernetes/client-go/blob/8f44946f6cbe967fbe2e2548e76987680a89428e/util/workqueue/default_rate_limiters.go#L136)

File: pkg/crcontrollers/reservation/reservationcontroller.go (hunk 4 above)
Comment (safronovD) [reply pkg/crcontrollers/reservation/reservationcontroller.go]: We have this check in the CRD. I wish to highlight here that `maxAttempts` is a natural number. Is it bad?

File: pkg/crcontrollers/reservation/reservationcontroller.go (hunk 5 above)
Comment (safronovD) [reply pkg/crcontrollers/reservation/reservationcontroller.go]: It is valid argument in RateLimiter, which means no timeout

**Review by mishoyama — APPROVED** [review]

**Review by rkiselenko — APPROVED** [review]

**Pull request merged**